test/pylib/resource_gather.py

```diff
@@ -0,0 +1,195 @@
+                    memory=int(memory_current.read()),
+                    timestamp=datetime.now()
+                )
+
+                self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+            # Control the frequency of updates, for example, every 2 seconds
+            await asyncio.sleep(1)
+
+    @staticmethod
+    def _parse_cpu_stat(file: TextIO, metrics: Metric) -> None:
+        # Map the values from cpu.state to the model. Keys are values from cpu.stat
+        stats = {'user_usec': 'user_sec', 'system_usec': 'system_sec', 'usage_usec': 'usage_sec'}
+        for line in file.readlines():
+            stat, value = line.split(' ')
+            if stat in stats.keys():
+                setattr(metrics, stats[stat], float(value) / 1_000_000)
+
+
+def get_resource_gather(is_switched_on, test, tmpdir: str) -> ResourceGather:
+    if is_switched_on:
+        return ResourceGatherOn(test, tmpdir)
+    else:
+        return ResourceGatherOff(test, tmpdir)
+
+
+def setup_cgroup(is_required: bool) -> None:
+    if is_required:
+        # check where the process is executed in podman or in docker
+        is_podman = os.access("/run/.containerenv", os.F_OK)
+        if is_podman:
+            subprocess.run(['sudo', 'chown', '-R', f"{getpass.getuser()}:{getpass.getuser()}", '/sys/fs/cgroup'],
+                           check=True)
+        else:
+            subprocess.run(
+                [
+                    "sudo",
+                    "mount",
+                    "-o",
+                    "remount,rw",
+                    "/sys/fs/cgroup",
+                ],
+                check=True,
+            )
+
+        for directory in [CGROUP_INITIAL, CGROUP_TESTS]:
+            directory.mkdir()
+
+        with open(CGROUP_INITIAL / 'cgroup.procs', "w") as f:
+            f.write(str(os.getpid()))
+
+        with open(CGROUP_INITIAL.parent / 'cgroup.controllers', "r") as f:
+            controllers = f.readline()
+        controllers = " ".join(map(lambda x: f"+{x}", controllers.split(" ")))
+
+        with open(CGROUP_INITIAL.parent / 'cgroup.subtree_control', "w") as f:
+            f.write(controllers)
+
+        with open(CGROUP_TESTS / 'cgroup.subtree_control', "w") as f:
+            f.write(controllers)
+
+
+async def monitor_resources(cancel_event: asyncio.Event, stop_event: asyncio.Event, tmpdir: Path) -> None:
+    """Continuously monitors CPU and memory utilization."""
+    sqlite_writer = SQLiteWriter(tmpdir / DEFAULT_DB_NAME)
+    while not cancel_event.is_set() and not stop_event.is_set():
+        timeline_record = SystemResourceMetric(
+            cpu=psutil.cpu_percent(interval=0.1),
+            memory=psutil.virtual_memory().percent,
+            timestamp=datetime.now()
+        )
+
+        sqlite_writer.write_row(timeline_record, SYSTEM_RESOURCE_METRICS_TABLE)
+
+        # Control the frequency of updates, for example, every 2 seconds
+        await asyncio.sleep(2)
```

Comment:
This is the second such coroutine, why two?

Reply:
One monitoring coroutine to gather system CPU and RAM utilization, the second one is to monitor the cgroup memory usage that is spawned for each test.

Reply:
I see. Do we need to keep system resource utilization in sqlite? It can't help future schedules.

Reply:
It can help to analyze the whole picture of where we're still not utilizing the RAM or CPU and where we can add additional thread.